runtime/moonbeam/src/xcm_config.rs

```diff
@@ -735,6 +762,273 @@ impl pallet_xcm_weight_trader::Config for Runtime {
+			gas_limit,
+			None,
+			None,
+			None,
+			Default::default(),
+			false,
+			false,
+			Some(weight_limit),
+			Some(0),
+			&<Runtime as pallet_evm::Config>::config(),
+		)
+		.map_err(|err| err.error)?;
+
+		ensure!(
+			matches!(
+				exec_info.exit_reason,
+				ExitReason::Succeed(ExitSucceed::Returned | ExitSucceed::Stopped)
+			),
+			DispatchError::Other("Contract transfer error")
+		);
+
+		// return value is true.
+		let mut bytes = [0u8; 32];
+		U256::from(1).to_big_endian(&mut bytes);
+
+		// Check return value to make sure not calling on empty contracts.
+		ensure!(
+			!exec_info.value.is_empty(),
+			DispatchError::Other("Contract is_nft_exists error")
+		);
+
+		Ok(exec_info.value == bytes)
+	}
+}
+
+type DerivativeNftDepositor =
+	UniqueInstancesDepositAdapter<AccountId, LocationToAccountId, EvmNftShim>;
+
+parameter_types! {
+	pub StashAccountId: AccountId = crate::Treasury::account_id();
+}
+
+type NftStash = SimpleStash<StashAccountId, EvmNftShim>;
+
+// This NFT engine which will:
+// 1. transfer the NFT to the stash on destroy
+// 2. on transfer it will just transfer
+// 3. transfer the NFT from the stash to the beneficiary on create
+//
+// We will use it as a parameter to the XCM adapter.
+type StashableNfts =
+	UniqueInstancesOps<RestoreOnCreate<NftStash>, EvmNftShim, StashOnDestroy<NftStash>>;
+
+fn try_get_full_derivative_nft_id(
+	asset_id: &xcm::latest::AssetId,
+	instance: &AssetInstance,
+) -> Result<FullNftId, DispatchError> {
+	let collection_info = crate::DerivativeNfts::get_derivative(asset_id)
+		.ok_or(pallet_derivatives::Error::<Runtime>::DerivativeNotFound)?;
+
+	let nft_id = match (instance, collection_info.instance_variant) {
+		(Index(id), crate::InstanceVariant::Index) => U256::from(*id),
+		(Array4(id), crate::InstanceVariant::Array4) => U256::from_big_endian(id),
+		(Array8(id), crate::InstanceVariant::Array8) => U256::from_big_endian(id),
+		(Array16(id), crate::InstanceVariant::Array16) => U256::from_big_endian(id),
+		(Array32(id), crate::InstanceVariant::Array32) => U256::from_big_endian(id),
+
+		_ => return Err(pallet_derivatives::Error::<Runtime>::DerivativeNotFound.into()),
+	};
+
+	let contract_addr = collection_info.contract_addr;
+
+	Ok((contract_addr, nft_id))
+}
+
+pub struct NftMatcher;
+impl MatchesInstance<FullNftId> for NftMatcher {
+	fn matches_instance(asset: &Asset) -> Result<FullNftId, MatchError> {
+		match (asset.id.0.unpack(), &asset.fun) {
+			(
+				(
+					0,
+					&[AccountKey20 {
+						key: contract_addr, ..
+					}],
+				),
+				&NonFungible(instance),
+			) if crate::DerivativeNfts::get_derivative(&asset.id).is_none() => {
+				let nft_id = match instance {
+					Undefined => return Err(MatchError::AssetNotHandled),
+					Index(id) => U256::from(id),
+					Array4(id) => U256::from_big_endian(&id),
+					Array8(id) => U256::from_big_endian(&id),
+					Array16(id) => U256::from_big_endian(&id),
+					Array32(id) => U256::from_big_endian(&id),
+				};
```

Comment:
```suggestion
				let nft_id = match instance {
					Array32(id) => U256::from_big_endian(&id),
					_ => return Err(MatchError::AssetNotHandled),
				};
```

В этой ветке, эти варианты говорят только о локальных NFT (созданных изначально на Мунбиме). Они все ERC-721, а значит имеют ID u256. Не думаю, что есть смысл для них принимать что-либо другое, кроме Array32.

Думаю, это имеет смысл исключительно для деривативов. Они могут иметь какую угодно идентификацию (например, ничего кроме `Index` на Юнике быть не может. А на другом чейне (или даже коллекции), например, может быть `Array4` и проч.). Для деривативов теперь уже всё учтено и этого достаточно

Reply:
Судя по комментариям в коде AssetInstance, Index это тоже подходящий вариант. И мне не очень понятно в чём минус поддержки остальных вариантов.